libraries/ms-common/inc/can_datagram.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "event_queue.h"
+#include "status.h"
+
+#define CAN_DATAGRAM_VERSION 1
+#define DATAGRAM_RX_BUFFER_LEN 256
+
+typedef StatusCode (*CanDatagramCb)(uint8_t *data, size_t len, bool start_message);
+
+typedef enum {
+  CAN_DATAGRAM_EVENT_RX = 0,
+  CAN_DATAGRAM_EVENT_TX,
+  CAN_DATAGRAM_EVENT_FAULT,
+  NUM_CAN_DATAGRAM_EVENTS,
+} CanDatagramCanEvent;
+
+typedef enum {
+  DATAGRAM_EVENT_PROTOCOL_VERSION = NUM_CAN_DATAGRAM_EVENTS + 1,
+  DATAGRAM_EVENT_CRC,
+  DATAGRAM_EVENT_DST_LEN,
+  DATAGRAM_EVENT_DT_TYPE,
+  DATAGRAM_EVENT_DST,
+  DATAGRAM_EVENT_DATA_LEN,
+  DATAGRAM_EVENT_DATA,
+  DATAGRAM_EVENT_COMPLETE,
+  DATAGRAM_EVENT_ERROR,
+  NUM_DATAGRAM_DIGEST_EVENTS,
+} CanDatagramEvent;
+
+typedef enum {
+  CAN_DATAGRAM_MODE_TX = 0,
+  CAN_DATAGRAM_MODE_RX,
+  NUM_CAN_DATAGRAM_MODES,
+} CanDatagramMode;
+
+typedef enum {
+  DATAGRAM_STATUS_OK = 0,
+  DATAGRAM_STATUS_COMPLETE,
+  DATAGRAM_STATUS_ERROR,
+} CanDatagramStatus;
+
+typedef enum {
+  // (TODO: SOFT_415) update datagram types
+  CAN_DATAGRAM_TYPE_A = 0,
+  CAN_DATAGRAM_TYPE_B,
+  CAN_DATAGRAM_TYPE_C,
+  CAN_DATAGRAM_TYPE_D,
+  NUM_CAN_DATAGRAM_TYPES,
+} CanDatagramType;
+
+typedef struct RxBufStore {
+  uint8_t data[8];
+  size_t len;
+} RxBufStore;
+
+typedef struct DatagramRxBuffer {
```

Comment:
Some of these structs are only used in the implementation I believe, see if you can move any of them to the C file so they aren't polluting client code namespaces.